File: frontend/src/types/workspace.ts
/**
 * Workspace 관련 wire types — Plan B foundation.
 *
 * backend `WorkspaceMeResponse` (com.ibizdrive.workspace.dto) 와 1:1.
 * spec docs/superpowers/specs/2026-05-09-team-centric-pivot-design.md §5.2.
 *
 * `kind`: backend `WorkspaceKind` enum 와 매칭 — 'department' | 'team' (정확 lower-case
 * 직렬화 — Plan A V12~V15 + Jackson default). 신규 'shared' 가상 종류는 frontend 전용
 * (UI에서만 의미 있음 — 사이드바 Section 3 + /shared/* 라우팅).
 */
export type WorkspaceKind = 'department' | 'team'

/** UI용 sidebar section 식별자 — `WorkspaceKind` 확장 + 'shared' 가상 섹션. */
export type SidebarSectionKind = 'department' | 'team' | 'shared'

export interface WorkspaceRef {
  kind: WorkspaceKind
  id: string
  name: string
  rootFolderId: string
}

/** GET /api/workspaces/me 응답. department 미배정 사용자는 null, teams는 항상 배열(0개 이상). */
export interface WorkspaceMeResponse {
  department: WorkspaceRef | null
  teams: WorkspaceRef[]
}

File: backend/src/main/java/com/ibizdrive/workspace/Workspace.java
package com.ibizdrive.workspace;

import java.util.UUID;

/**
 * Polymorphic facade over {@code Department} and {@code Team} for permission evaluation
 * and sidebar API (spec docs/superpowers/specs/2026-05-09-team-centric-pivot-design.md §1).
 *
 * <p>Callers obtain a {@code Workspace} via {@link DepartmentWorkspace} or {@link TeamWorkspace}
 * adapter records. No factory method is provided — adapters are constructed directly (YAGNI).
 *
 * <p>This interface is intentionally non-sealed. The two known implementations cover the current
 * scope; a sealed hierarchy would require a {@code permits} clause that couples this interface to
 * concrete adapters (KISS).
 */
public interface Workspace {

    /**
     * Discriminator for permission evaluation and routing.
     *
     * @return {@link WorkspaceKind#DEPARTMENT} or {@link WorkspaceKind#TEAM} — never null
     */
    WorkspaceKind kind();

    /**
     * Scope identifier used in {@code folders.scope_id} and {@code permissions.subject_id}
     * (spec §1). Matches the underlying entity's primary key.
     *
     * @return non-null UUID identifying this workspace scope
     */
    UUID id();

    /**
     * Display name of the workspace as stored on the underlying entity.
     *
     * @return non-null, non-blank display name
     */
    String name();

    /**
     * Root folder UUID for this workspace scope. May be {@code null} for departments whose
     * root folder has not yet been created (Plan A2 backfill).
     *
     * @return root folder UUID, or {@code null} if not yet attached
     */
    UUID rootFolderId();

    /**
     * Whether the workspace is currently active (not soft-deleted or archived).
     * For departments this mirrors {@code deletedAt == null}; for teams it mirrors
     * {@code archivedAt == null}.
     *
     * @return {@code true} if the workspace is active
     */
    boolean isActive();
}
